File: bin/m_put.py
#!/usr/bin/env python
# 2019/04/23: pass through rsync arguments
# when?: converted to python

import read_data_pm as readpm

import argparse
import os
import subprocess

parser = argparse.ArgumentParser(
        description='put files into clusters. can additionally accept rsync arguments')
# for individual files/dir
parser.add_argument('--files', '-f', nargs='*')
# optional info if not supplied by *.jss, UserName, or Jobdir
parser.add_argument('--cluster', '-c')
parser.add_argument('--username', '-u')
parser.add_argument('--destination', '-s')
parser.add_argument('--no-delete', '-n', action='store_true')
args, rsync_args = parser.parse_known_args()

cname = args.cluster if args.cluster else readpm.specify_cluster()
uname = args.username if args.username else readpm.get_uname(cname)

datapm = readpm.read_perl_module_hashes(readpm.DATAPM)
c1 = datapm['clusters']['unified']
c2 = datapm['clusters'][cname]

def build_command(key, arg):
    # output: [ a b1 a b2 ... ] from f(a, [ b1 b2 ... ])
    return [ x for y in [[key, x] for x in arg] for x in y ]

# add includes and excludes, starting with data.pm defaults
command = list(filter(None, ['rsync', '-avz', '--progress', c2['rsync_option']]))
command.extend(build_command('--exclude', readpm.exclude))
command.extend(build_command('--include', readpm.include))

# rsync pass-through
command.extend(rsync_args)

path = args.destination if args.destination else readpm.c2path(c1, c2, uname, cname)
files = args.files if args.files else [ os.path.join(os.getcwd(), '') ]
command.extend([ x for x in files ])
command.extend([ '{}:{}'.format(readpm.c2ip(c2, uname), os.path.join(path, '')) ])

mkdir = list(filter(None, [ 'ssh', c2['ssh_option'], readpm.c2ip(c2, uname), 'mkdir -p {}'.format(path) ]))

if args.no_delete:
    command = [ x for x in command if x != '--delete' ]
if not '--dry-run' in rsync_args:
    print(subprocess.list2cmdline(mkdir))
    subprocess.call(mkdir)
print(subprocess.list2cmdline(command))
subprocess.call(command)

File: bin/m_put.pl
#!/usr/bin/perl
#############################################
#
#   25Oct18  ;  Implement dry run option
#   25Apr18  ;  Allow to specify custom destination, general cleanup
#               Add include and exclude
#   05Feb18  ;  Add 'help' option
#   03Feb18  ;  Finished new parsing
#   30Jan18  ;  update: not a fan of how both scripts work (aput mindlessly use *jss where it does not exists, data::specifyCluster(), put does not check for arguments
#               solution: make it an independent script that works more like rsync but fitted for clusters and use data.pm's data
#   30Jan18  ;  This is more like a subsitute to aput.pl (but with multiple files capability)
#               Therefore make it more like aput (name, behaviour, and minimal diff)
#               Except for last part, where put.pl's rsync is superior
#   11Jan18  ;  Forked by genki
#               Added function to transfer individual files when supplied
#   24Aug15  ;  First composed by ichibha
#
############################################
#TODO: pass rsync options directly to rsync

use Cwd;
use lib $ENV{CLUSTER_SCRIPTS_TOP}; 
use data;
use Getopt::Long;
use File::Basename;
use File::Spec::Functions 'catfile';
#use warnings;
$0=basename($0);

my $delete = 0;
my $help = 0;
my $dry_run = 0;
GetOptions( 'cluster=s' => \my $cname,
            'username=s' => \my $uname,
            'exclude=s' => \my @excludes,
            'include=s' => \my @includes,
            'destination=s' => \my $dest,
            'help' => sub { $help = 1 },
            'delete' => sub { $delete = 1 },
            'dry-run' => sub { $dry_run = 1},
) or die "$0: Invalid arguments\n"; die "Usage: $0 [OPTION]... [FILE]...\n" if $help;
$cname = defined $cname ? $cname : &data::specifyCluster();
$cname = $cname eq '' ? &data::inputCluster() : $cname;
$rsyncopt = $delete ? '--delete' : '';
if ($uname eq '') {
  &data::getUserName();
  $uname = $data::userName;
}

my $c1dir = $data::clusters{unified}{base_dir};
my $c2dir = $data::clusters{$cname}{base_dir};
my $c2ip   = $data::clusters{$cname}{ip_address}; $c2ip =~ s/USER/$uname/;
my $c2rsyncopt = $data::clusters{$cname}{rsync_option};
my $c2sshopt  = $data::clusters{$cname}{ssh_option};
my $cwd = Cwd::getcwd().'/';

my $path;
if (defined $dest) {
  $path = $dest;
} else {
  my $base = $cwd;
  $base =~ s/$c1dir//;
#  $base =~ s/$c1dir2//;
  $path = "$c2dir$base";
  $path =~ s/USER/$uname/;
}

my @files = @ARGV;
my $files = @files ? '' : $cwd;
foreach my $filename (@files) {
  die "$0: $filename: No such file\n" unless -e $filename;
  $files .= " ".catfile($cwd,$filename);
}

my $rsyncin = '';
my $rsyncex = '';
foreach my $include (@includes) {
  $rsyncin = $rsyncin." --include=\"$include\" ";
}
foreach my $exclude (@excludes) {
  $rsyncex = $rsyncex." --exclude=\"$exclude\" ";
}

my $ssh = "ssh $c2sshopt $c2ip 'mkdir -p $path'";
my $rsync = "rsync -avz --progress $data::exclude $data::include $rsyncin $rsyncex $rsyncopt $c2rsyncopt $files $c2ip\:$path";

print $ssh."\n";
print $rsync."\n";
unless ($dry_run) {
    system $ssh;
    system $rsync;
}
